Code/GraphMol/MolOps.cpp

```diff
@@ -881,5 +890,184 @@ unsigned getNumAtomsWithDistinctProperty(const ROMol &mol, std::string prop) {
+// atoms passed in, with a dative bond from it to the metal atom.
+void addDativeBond(RWMol &mol, unsigned int metalIdx,
+                   std::vector<unsigned int> hapticAtoms) {
+  // This rigmarole is required to get a * in the V3000 file as the symbol
+  // for the atom.
+  auto dummyAt = new QueryAtom(0);
+  auto *res = new ATOM_NULL_QUERY;
+  res->setDataFunc(nullDataFun);
+  res->setMatchFunc(nullQueryFun);
+  res->setDescription("AtomNull");
+  dummyAt->setQuery(res);
+  unsigned int dummyIdx = mol.addAtom(dummyAt);
+  RDGeom::Point3D dummyPos;
+  std::ostringstream oss;
+  for (auto ha : hapticAtoms) {
+    auto haPos = mol.getConformer().getAtomPos(ha);
+    dummyPos.x += haPos.x;
+    dummyPos.y += haPos.y;
+    oss << ha << " ";
+  }
+  dummyPos.x /= hapticAtoms.size();
+  dummyPos.y /= hapticAtoms.size();
+  mol.getConformer().setAtomPos(dummyIdx, dummyPos);
+  unsigned int numbonds = mol.addBond(dummyIdx, metalIdx, Bond::DATIVE);
+  auto bond = mol.getBondWithIdx(numbonds - 1);
+  std::string endpts = "(" + oss.str();
+  if (endpts[endpts.length() - 1] == ' ') {
+    endpts = endpts.substr(0, endpts.length() - 1);
+  }
+  endpts += ")";
+  bond->setProp(common_properties::_MolFileBondEndPts, endpts);
+}
+}  // namespace
+
+void dativeBondsToHaptic(RWMol &mol) {
+  // first collect all the atoms that have a dative bond to them.
+  std::map<unsigned int, std::vector<unsigned int>> dativeAtoms;
+  for (const auto &b : mol.bonds()) {
+    if (b->getBondType() == Bond::DATIVE) {
+      auto ins = dativeAtoms.find(b->getEndAtomIdx());
+      if (ins == dativeAtoms.end()) {
+        dativeAtoms.insert(
+            std::make_pair(b->getEndAtomIdx(),
+                           std::vector<unsigned int>{b->getBeginAtomIdx()}));
+      } else {
+        ins->second.push_back(b->getBeginAtomIdx());
+      }
+    }
+  }
+
+  for (auto &dativeSet : dativeAtoms) {
+    auto contigAtoms = contiguousAtoms(mol, dativeSet.second);
+    for (const auto &ca : contigAtoms) {
+      addDativeBond(mol, dativeSet.first, ca);
+      for (auto cat : ca) {
+        mol.removeBond(dativeSet.first, cat);
+      }
+    }
+  }
+}
+
+namespace details {
+std::vector<int> hapticBondEndpoints(const Bond *bond) {
+  // This would ideally use ParseV3000Array but I'm buggered if I can get
+  // the linker to find it.  The issue, I think, is that it's in the
+  // FileParsers library which is built after GraphMol so not available
+  // to link in.  It can't be built first because it needs GraphMol.
+  //      std::vector<unsigned int> oats =
+  //          RDKit::SGroupParsing::ParseV3000Array<unsigned int>(endpts);
+  // Returns the atom indices i.e. subtracts 1 from the numbers in the prop.
+  std::vector<int> oats;
+  std::string endpts;
+  if (bond->getPropIfPresent(common_properties::_MolFileBondEndPts, endpts)) {
+    if ('(' == endpts.front() && ')' == endpts.back()) {
+      endpts = endpts.substr(1, endpts.length() - 2);
```

Comment:
since we are now in C++17, this block can be made more efficient:
```
      std::string_view v = endpts;
      v = v.substr(1, endpts.length() - 2);
      boost::char_separator<char> sep(" ");
      boost::tokenizer<boost::char_separator<char>> tokens(v, sep);
```

Reply:
I can't make string_view and the boost::tokenizer play together nicely.  Since this is going to go away when the end points are ints, I vote for leaving it as is for now.